yew/src/html/mod.rs

```diff
@@ -31,12 +32,50 @@ cfg_if! {
 /// This type indicates that component should be rendered again.
 pub type ShouldRender = bool;
 
+/// Wrapper for legacy components
+#[derive(Debug)]
+pub struct Legacy<T: LegacyComponent>(T);
+
+impl<T: LegacyComponent> yew::component::Component for Legacy<T> {
+    type Message = T::Message;
+    type Properties = T::Properties;
+
+    fn create(ctx: &Scope<Self>) -> Self {
+        let props: &Self::Properties = &ctx.props;
+        Self(T::create(props.clone(), ctx.clone()))
+    }
+
+    fn changed(&mut self, _ctx: &Context<Self>, new_props: &Self::Properties) -> ShouldRender {
+        self.0.change(new_props.clone())
+    }
+
+    fn update(&mut self, _ctx: &Context<Self>, msg: Self::Message) -> ShouldRender {
+        self.0.update(msg)
+    }
+
+    fn view(&self, _ctx: &Context<Self>) -> Html {
+        self.0.view()
+    }
+
+    fn rendered(&mut self, _ctx: &Context<Self>, first_render: bool) {
+        self.0.rendered(first_render)
+    }
+
+    fn destroy(&mut self, _ctx: &Context<Self>) {
+        self.0.destroy()
+    }
+}
+
 /// Components are the basic building blocks of the UI in a Yew app. Each Component
 /// chooses how to display itself using received props and self-managed state.
 /// Components can be dynamic and interactive by declaring messages that are
 /// triggered and handled asynchronously. This async update mechanism is inspired by
 /// Elm and the actor model used in the Actix framework.
-pub trait Component: Sized + 'static {
+#[deprecated(
+    since = "0.18.0",
+    note = "Please switch to the yew::component::Component trait"
+)]
+pub trait LegacyComponent: Sized + 'static {
```

Comment:
This was the big dilemma I had when making these changes. Ideal situation would be:
1) Create new Component interface which works well with yew-macro
2) Deprecate existing Component interface but have it still be compatible with yew-macro

But these two are conflicting at the macro level. I decided to prioritize the new interface experience over having a nice deprecated experience but try my best to minimize changes needed to switch to the legacy component interface.

I opted to rename to `LegacyComponent` to make it obvious to devs which components have been upgraded but I don't feel too strongly about that. The main annoying thing is the macro changes to wrap all the `html!` components with `Legacy<..>`. Perhaps it is too annoying. I really care about this upgrade being smooth because upgrading to the new `Component` is tedious.

Call me crazy, but @siku2 what do you think about adding a new macro? It would act the same as `html!` but would handle the new `Component` interface. I'm thinking it could be called `vdom!`

Reply:
I feel very conflicted about this. I very much understand the need to make the transition as seamless as possible but we might be worrying a bit too much about this given that Yew is officially still at major version zero (i.e. initial development).

To quote [semver.org](https://semver.org/#how-do-i-know-when-to-release-100):
> If you’re worrying a lot about backwards compatibility, you should probably already be 1.0.0.

Adding a new macro is definitely too far in my opinion. If you think the migration to the new component trait is too tedious, I'm fine with keeping `LegacyComponent` around until the following release. We can deal with the `Legacy<_>` wrapper by creating a type alias `type MyActualComponent = Legacy<MyComponentImplementation>`.

The migration guide looks something like this:

1. Replace `impl Component` with `impl LegacyComponent`
2. Rename your component `Foo` to something like `FooComp` locally
3. Create a type alias `pub type Foo = Legacy<FooComp>;`
4. Use the `Foo` type instead of the `FooComp` struct

Personally I wouldn't even consider switching to `LegacyComponent`, if I have to do some modifications in each file anyway, I might as well upgrade to the new `Component` trait.